crates/external_services/src/grpc_client/dynamic_routing.rs

```diff
@@ -0,0 +1,265 @@
+        field: String,
+    },
+    /// Error from Dynamic Routing Server
+    #[error("Error from Dynamic Routing Server : {0}")]
+    SuccessRateBasedRoutingFailure(String),
+}
+
+/// Type that consists of all the services provided by the client
+#[derive(Debug, Clone)]
+pub struct RoutingStrategy {
+    /// success rate service for Dynamic Routing
+    pub success_rate_client: Option<SuccessRateCalculatorClient<Channel>>,
+}
+
+/// Contains the Dynamic Routing Client Config
+#[derive(Debug, Clone, serde::Deserialize, serde::Serialize, Default)]
+#[serde(untagged)]
+pub enum DynamicRoutingClientConfig {
+    /// If the dynamic routing client config has been enabled
+    Enabled {
+        /// The host for the client
+        host: String,
+        /// The port of the client
+        port: u16,
+    },
+    #[default]
+    /// If the dynamic routing client config has been disabled
+    Disabled,
+}
+
+impl DynamicRoutingClientConfig {
+    /// establish connection with the server
+    pub async fn get_dynamic_routing_connection(
+        self,
+    ) -> Result<RoutingStrategy, Box<dyn std::error::Error>> {
+        let success_rate_client = match self {
+            Self::Enabled { host, port } => {
+                let uri = format!("http://{}:{}", host, port);
+                let channel = tonic::transport::Endpoint::new(uri)?.connect().await?;
+                Some(SuccessRateCalculatorClient::new(channel))
+            }
+            Self::Disabled => None,
+        };
+        Ok(RoutingStrategy {
+            success_rate_client,
+        })
+    }
+}
+
+/// The trait Success Based Dynamic Routing would have the functions required to support the calculation and updation window
+#[async_trait::async_trait]
+pub trait SuccessBasedDynamicRouting: dyn_clone::DynClone + Send + Sync {
+    /// To calculate the success rate for the list of chosen connectors
+    async fn calculate_success_rate(
+        &self,
+        id: String,
+        success_rate_based_config: SuccessBasedRoutingConfig,
+        label_input: Vec<RoutableConnectorChoice>,
+    ) -> DynamicRoutingResult<CalSuccessRateResponse>;
+    /// To update the success rate with the given label
+    async fn update_success_rate(
+        &self,
+        id: String,
+        success_rate_based_config: SuccessBasedRoutingConfig,
+        response: Vec<RoutableConnectorChoiceWithStatus>,
+    ) -> DynamicRoutingResult<UpdateSuccessRateWindowResponse>;
+}
+
+#[async_trait::async_trait]
+impl SuccessBasedDynamicRouting for SuccessRateCalculatorClient<Channel> {
+    async fn calculate_success_rate(
+        &self,
+        id: String,
+        success_rate_based_config: SuccessBasedRoutingConfig,
+        label_input: Vec<RoutableConnectorChoice>,
+    ) -> DynamicRoutingResult<CalSuccessRateResponse> {
+        let params = success_rate_based_config
+            .params
+            .map(|vec| {
+                vec.into_iter().fold(String::new(), |mut acc_vec, params| {
```

Comment:
`acc_vec` is a `String` here and not a `Vec`, isn't it? Can we consider using a different variable name?